### DISCUSSION TITLE
Option to use streamlink-plugin ttvlol to get rid of ads

### DISCUSSION THREAD

**Discussion by makiele** [MrBrax/LiveStreamDVR#425] [Ideas]
I am currently experimenting with [ streamlink-ttvlol-proxy ](https://github.com/2bc4/streamlink-ttvlol) to get ride of ads when using streamlink. At the moment, I am testing it with the streamlink-twitch-gui on my pc. 

What would be the best way to get this working in LiveStreamDVR? 
Would I just replace the twitch-plugin in `/usr/local/lib/python3.9/dist-packages/streamlink/plugins` in the docker-container?
And how would I add flags like `--twitch-proxy-playlist=https://api.ttv.lol ` needed for the proxy to the streamlink command?

**Comment by MrBrax** [comment]
you could test it by changing the dockerfile to add the plugin yeah (maybe even copy it inside somehow while it's running), and arguments are in this file:

https://github.com/MrBrax/LiveStreamDVR/blob/31a6e3fa23a395db0090ae26ace2f9d6acd8c15f/server/src/Core/Providers/Twitch/TwitchAutomator.ts#L382

it does however need to get rebuilt, so a `yarn run build` in the `/usr/local/share/twitchautomator/` folder is required.

i'll keep this idea in mind if i get back into this in any reasonable amount of time.

**Comment by makiele** [comment]
ok, did the following:

- `docker exec -it CONTAINER-ID bash` to connect to the container
- `mv /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch.py /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch_old.py ` to make a backup of the original twitch-plugin
- `curl -L -o /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch.py 'https://github.com/2bc4/streamlink-ttvlol/releases/latest/download/twitch.py'` to download the ttvlol-plugin to its location
- open `TwitchAutomator.ts` with `nano /usr/local/share/twitchautomator/server/src/Core/Providers/Twitch/TwitchAutomator.ts`
- add the necessary arguments after line 382:` cmd.push("--twitch-proxy-playlist=https://eu.luminous.dev,https://eu2.luminous.dev,https://lb-eu.perfprod.com,https://api.ttv.lol");`

- save everything

the only thing I don't get as a docker-noob - where do I have to run `yarn run build`? `/usr/local/share/twitchautomator/` does not have a packages.json and `/usr/local/share/twitchautomator/server` has one, but fails with `yarn run build`

```
/usr/local/share/twitchautomator/server/.pnp.cjs:17586
  return Object.defineProperties(new Error(message), {
                                 ^

Error: Required package missing from disk. If you keep your packages inside your repository then restarting the Node process may be enough. Otherwise, try to run an install first.

```

after restarting the docker container, the plugin is recongized by LiveStreamDVR according to the logs 
```
[plugins.twitch][info] Streamlink Twitch plugin (2bc4 fork).
[plugins.twitch][info] Please report issues to https://github.com/2bc4/streamlink-ttvlol/issues
```
but the new arguments are missing (as i don't know what exactly I should build):
```
[cli][debug]  --default-stream=['best']
[cli][debug]  --retry-streams=10.0
[cli][debug]  --retry-max=5
[cli][debug]  --hls-live-edge=99999
[cli][debug]  --hls-segment-threads=5
[cli][debug]  --hls-segment-timeout=200
[cli][debug]  --hls-timeout=200
[cli][debug]  --ffmpeg-fout=mpegts
[cli][debug]  --twitch-disable-ads=True
[cli][debug]  --twitch-disable-hosting=True
[cli][debug]  --twitch-disable-reruns=True
```

**Comment by MrBrax** [comment]
ah right that error makes sense, the dockerfile cleans out the node modules, maybe try `yarn install` also? and yeah it's in the server folder

**Reply by makiele** [reply]
yes, that did it - thank you! will try it out and report back if the amount of ads gets reduced.

thanks again for the quick and helpful support!

**Comment by makiele** [comment]
for anyone else interested, here a quick summary how to add the plugin to LiveStreamDVR:

- `docker exec -it CONTAINER-ID bash` to connect to the container
- `mv /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch.py /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch_old.py ` to make a backup of the original twitch-plugin
- `curl -L -o /usr/local/lib/python3.9/dist-packages/streamlink/plugins/twitch.py 'https://github.com/2bc4/streamlink-ttvlol/releases/latest/download/twitch.py'` to download the ttvlol-plugin to its location
- edit`TwitchAutomator.ts` with `nano /usr/local/share/twitchautomator/server/src/Core/Providers/Twitch/TwitchAutomator.ts`
- add the necessary arguments after line 382:` cmd.push("--twitch-proxy-playlist=https://eu.luminous.dev,https://eu2.luminous.dev,https://lb-eu.perfprod.com,https://api.ttv.lol");` (see [readme](https://github.com/2bc4/streamlink-ttvlol#flags) for every possible arguments and [ttv-lol-pro](https://github.com/younesaassila/ttv-lol-pro/discussions/37) for possible proxy servers)

- save everything
- go to /`usr/local/share/twitchautomator/server` and run `yarn install`, then `yarn run build`
- restart docker container

**Comment by MrBrax** [comment]
are the proxy services closed source? i wonder if it could just be ran inside the container locally

**Reply by makiele** [reply]
i am not entirely familiar with the whole process in the background but as far as i understood, these servers are just used to connect to the hola network, request the playlist from a non-ad country and return the url to the user.

[Luminous TTV](https://github.com/AlyoshaVasilieva/luminous-ttv) has a write up in their readme

**Comment by makiele** [comment]
### small update after 3 days of using the plugin with LiveStreamDVR:

about 80% of my recorded streams have seen a massive reduction in ads received. 

One streamer for example dropped from 20 minutes stream pauses per stream to 1 minute for one stream and to no ads for another stream (each of the example streams with a duration of 6 hours). Other streams/streamers dropped to 0 ads when there were 5 to 15 minutes of ads before.

A few of the streams saw a minimal reduction, which I guess is due to the proxies used. Nevertheless I am still experimenting with all the options for the plugin and the different proxies, so I think there is still room for improvement with some tweaks

### Example for one streamer:
**without proxy-plugin**
![image](https://user-images.githubusercontent.com/62406247/228882626-fcf37e01-16ab-4479-b056-f53a217a614d.png)
![image](https://user-images.githubusercontent.com/62406247/228882711-7676f6a9-7a96-49b2-b993-2c2b24c7a687.png)

**with proxy-plugin**
![image](https://user-images.githubusercontent.com/62406247/228882797-09ed6eba-2df8-4a03-87a0-90a48ecd112f.png)
![image](https://user-images.githubusercontent.com/62406247/228882860-ed75b74a-55d6-4d28-b93e-1dd7d2f6850c.png)

**Comment by MrBrax** [comment]
very interesting, about a minute was the previous delay because of the preroll so it's very close now.

i'm just wondering about the legitimacy of including this, like there must be a reason for streamlink not including it

**Reply by makiele** [reply]
hard to tell - may be due to a variety of concerns:

- they want to record the content as "pure" as possible - in the end, using a proxy will bring a variety of experiences, which may often work, but sometimes just won't. 
- additionally, they would be dependent on proxy-providers, if one stops working, a lot of users would probably blame them.
- also, adding the proxy-plugin to streamlink is pretty easy - they may think that if someone really wants it, adding the plugin by themselves isn't that hard for end users. 
- they did mention proxy servers in their[ ads issue](https://github.com/streamlink/streamlink/issues/4949)

in the end, I feel like they don't want to be trapped in a constant cat and mouse game with users banging on their door to fix stuff they just can't fix

**Comment by MrBrax** [comment]
i wonder, will putting that python file in the install overwrite anything? like can it be enabled or disabled at will

**Comment by MrBrax** [comment]
i pushed a possible experimental solution for this, a few new config options are available on dev

**Reply by makiele** [reply]
damn, thanks for making it possible. will look into it

**Comment by MrBrax** [comment]
can confirm now that it works 'officially'

**Comment by makiele** [comment]
just a heads up - twitch seems to have changed something in the past days. ttvlol currently has a [v2 beta](https://github.com/younesaassila/ttv-lol-pro/releases/tag/v2.0.0-beta.3) in testing (also see[ #131](https://github.com/younesaassila/ttv-lol-pro/discussions/131)) that seems to get adblocking working again. the streamlink-plugin hasn't been updated yet.
